runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/StateMultiplexingGroupByKey.java

```diff
@@ -0,0 +1,316 @@
+    KvCoder<K, V> inputCoder = getInputKvCoder(input.getCoder());
+
+    // Ensure that the output coder key and value types aren't different.
+    Coder<?> outputCoder = Iterables.getOnlyElement(outputs.values()).getCoder();
+    KvCoder<?, ?> expectedOutputCoder = getOutputKvCoder(inputCoder);
+    if (!expectedOutputCoder.equals(outputCoder)) {
+      throw new IllegalStateException(
+          String.format(
+              "the GroupByKey requires its output coder to be %s but found %s.",
+              expectedOutputCoder, outputCoder));
+    }
+  }
+
+  // Note that Never trigger finishes *at* GC time so it is OK, and
+  // AfterWatermark.fromEndOfWindow() finishes at end-of-window time so it is
+  // OK if there is no allowed lateness.
+  private static boolean triggerIsSafe(WindowingStrategy<?, ?> windowingStrategy) {
+    if (!windowingStrategy.getTrigger().mayFinish()) {
+      return true;
+    }
+
+    if (windowingStrategy.getTrigger() instanceof NeverTrigger) {
+      return true;
+    }
+
+    if (windowingStrategy.getTrigger() instanceof FromEndOfWindow
+        && windowingStrategy.getAllowedLateness().getMillis() == 0) {
+      return true;
+    }
+
+    if (windowingStrategy.getTrigger() instanceof AfterWatermarkEarlyAndLate
+        && windowingStrategy.getAllowedLateness().getMillis() == 0) {
+      return true;
+    }
+
+    if (windowingStrategy.getTrigger() instanceof AfterWatermarkEarlyAndLate
+        && ((AfterWatermarkEarlyAndLate) windowingStrategy.getTrigger()).getLateTrigger() != null) {
+      return true;
+    }
+
+    return false;
+  }
+
+  @Override
+  public PCollection<KV<K, Iterable<V>>> expand(PCollection<KV<K, V>> input) {
+    applicableTo(input);
+    // Verify that the input Coder<KV<K, V>> is a KvCoder<K, V>, and that
+    // the key coder is deterministic.
+    Coder<K> keyCoder = getKeyCoder(input.getCoder());
+    Coder<V> valueCoder = getInputValueCoder(input.getCoder());
+    KvCoder<K, Iterable<V>> outputKvCoder = getOutputKvCoder(input.getCoder());
+
+    try {
+      keyCoder.verifyDeterministic();
+    } catch (NonDeterministicException e) {
+      throw new IllegalStateException("the keyCoder of a GroupByKey must be deterministic", e);
+    }
+    Preconditions.checkArgument(numShardingKeys > 0);
+    final TupleTag<KV<ByteString, V>> largeKeys = new TupleTag<KV<ByteString, V>>() {};
+    final TupleTag<KV<ByteString, V>> smallKeys = new TupleTag<KV<ByteString, V>>() {};
+    WindowingStrategy<?, ?> originalWindowingStrategy = input.getWindowingStrategy();
+
+    PCollectionTuple mapKeysToBytes =
+        input.apply(
+            "MapKeysToBytes",
+            ParDo.of(
+                    new DoFn<KV<K, V>, KV<ByteString, V>>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c) {
+                        KV<K, V> kv = c.element();
+                        Output output = ByteString.newOutput();
```

Comment:
ByteStringOutputStream is optimized for this

Reply:
done